Update the workflow for the tests

lionelkusch opened the pull request:
For managing the workflow, I create a file `main_workflow.yml`. This file reuse the other workflow and it contains the condition between them.
I include the following conditions:
- when the linter is falling, the tests and the documentation are not run
- when the tests ares failing, the documentation is node made.

For the tests, I try to find the best way to get the information is to display at the right place. 
I find 3 ways to display information in github:
- report xml with [junit](https://docs.pytest.org/en/stable/reference/reference.html#command-line-flags) (--junitxml) and [test-reporter](https://github.com/marketplace/actions/test-reporter)
- report json with [ctrf](https://pypi.org/project/pytest-json-ctrf/) and [test-reporter](https://github.com/ctrf-io/github-test-reporter) 
- report mardown with [pytest-md-report](https://github.com/thombashi/pytest-md-report) and [push mardown](https://github.com/thombashi/pytest-md-report#add-report-to-pull-requests) in summary in PR.

I chose the last one which seems for me the most robust and the most flexible.

I add also add the generation of html reports and a report for the duration of the tests. 
For the HMTL report, there are 2 interesting plugging [pytest-reporter-html1](https://pypi.org/project/pytest-reporter-html1/) and [pytest-html](https://pypi.org/project/pytest-html). I like it more the pytest-html.
For the duration, I am not sure that the report, for the duration, is interesting to keep it.
I split a test in two to reduce the duration of 1 test because I include a limitation of the time for the tests and the overall tests.

Remi-Gau commented on .github/workflows/call_black.yml:
```diff
@@ -0,0 +1,15 @@
+# For more information see: https://black.readthedocs.io/en/stable/integrations/github_actions.html
+name: Lint
+
+on: 
+  # Allows you to run this workflow from another workflow
+  workflow_call
```
oh I did not know this was doable: this should help with making reusable workflow component

Remi-Gau commented on .github/workflows/call_test_package.yml:
```diff
@@ -0,0 +1,142 @@
+      REPORT_NAME: ${{ matrix.python-version }}-os-${{ matrix.os }}
+
+    steps:
+    # Checkout to the branch for getting the ritgh code
```
```suggestion
    # Checkout to the branch for getting the right code
```

Remi-Gau commented on .github/workflows/call_test_package.yml:
```diff
@@ -0,0 +1,142 @@
+    # option for duration report: --pytest-durations= number of second --pytest-resultlog=log-...
+    - name: Generate coverage report
+      run: |
+        python -m pytest\
+         --cov=. --cov-report=xml:cov/cov-${{env.REPORT_NAME}}.xml\
+         --md-report --md-report-output=test-results-${{env.REPORT_NAME}}.md\
+         --self-contained-html --html=report-${{env.REPORT_NAME}}.html\
+         --pytest-durations=10 --pytest-resultlog=log-${{env.REPORT_NAME}}.txt
```
see if you can move some of those to the pytest config in pyproject.toml:
- this will make sure they run all the time
- this will make the CI set up easier to read

or maybe keep them there for now and then move them your tox command if / when you set up tox

lionelkusch replied:
I added these options only here because there are the generation of reports which I don't need when I run the test on my computer.

Remi-Gau commented on .github/workflows/call_test_package.yml:
```diff
@@ -0,0 +1,142 @@
+      id: html_report
+      uses: actions/upload-artifact@v4
+      with:
+        name: pytest-results-html-${{env.REPORT_NAME}}
```
I think you may get a name clash error because the different runs of your testing matrix will have the same report filename

lionelkusch replied:
No, because ${{env.REPORT_NAME}} take into consideration of the matrix.
This is the cause for having jobs, after which will merge the different reports.

Remi-Gau commented on .github/workflows/call_test_package.yml:
```diff
@@ -0,0 +1,142 @@
+    runs-on: ubuntu-latest
+    if: ${{ failure() }}
+    steps:
+    # get report from previous job 
+    - name: Get data
+      uses: actions/download-artifact@v4
+      with:
+        pattern: pytest-results-md-*-os-*
+    # create a comment in the PR with the failling tests
+    # for more details: https://github.com/marocchino/sticky-pull-request-comment
+    - name: Render the report to the PR when tests fail
+      uses: marocchino/sticky-pull-request-comment@v2
+      with:
+        header: test-report
+        recreate: true
+        path: ./pytest-results-md-3.13-os-macos-latest/test-results-3.13-os-macos-latest.md     #CHECK: need to be updated when new version are available
```
yeah you may want to have env variable to keep track of which OS and python version you want to use as 'source' to generate the report

lionelkusch replied:
In this case, I chose to summit only one report because it's the first to run, i.e. the first to fail, and to avoid having too many notifications.

lionelkusch replied:
I finally put all the reports on circleci.

Remi-Gau commented on .github/workflows/call_test_package.yml:
```diff
@@ -0,0 +1,142 @@
+          echo "</details>" >> $GITHUB_STEP_SUMMARY
+        done
+
+  merge_upload_coverage_data:
```
no sure this is necessary: I think codecov does that for you

or least I don't ever remember having to 'manually' merge xml before sending them to codecov

Remi-Gau commented on .github/workflows/main_workflow.yml:
```diff
@@ -0,0 +1,40 @@
+# Management of the action do made during a pull request
+
+name: 'Main_workflow'
+on:
+  pull_request:
+    branches: main
+  push:
+    branches: main
+
+# Cancel existing runs 
+concurrency:
+  group: ${{ github.workflow }}-${{ github.ref }}
+  cancel-in-progress: true
+
+jobs:
+  linter:
+    name: Linter
+    uses: ./.github/workflows/call_black.yml
+  
+  tests:
+    name: tests
+    needs: [linter]
+    uses: ./.github/workflows/call_test_package.yml
+
+  make_documentation:
+    name: Make_documentation
+    needs: [linter, tests]
+    uses: ./.github/workflows/call_build_docs.yml
```
really nice that github allows to compose workflow like this: is this a new thing? or have I been living under a rock?

lionelkusch replied:
It's a feature which was introduced in 2020.
For me, it seems that the feature is used generally by people but I like it because I feel that this gives a better way to manage the workflow.

jpaillard reviewed (commented):
Do you think #175 be addressed here or in a follow-up PR?

lionelkusch commented:
> Do you think #175 be addressed here or in a follow-up PR?

For the documentation, I prefer to include in other PR because it will require including the creation of another repository or a new branch.
I would like to address it after the sprint because I am assuming that the example will be faster to run.

lionelkusch commented:
For desirable behaviour, see the [PR 4](https://github.com/lionelkusch/hidimstat/pull/4) on my fork.

bthirion commented on test/test_ensemble_clustered_inference.py:
```diff
@@ -76,6 +76,8 @@ def test_ensemble_clustered_inference():
     )
 
+
+def test_ensemble_clustered_inference_senario_2():
```
Maybe add a comment line too.

bthirion reviewed (commented):
LGMT overall.

lionelkusch commented:
I updated the workflow.
![image](https://github.com/user-attachments/assets/8f5a1a01-ed1b-408f-a036-2078d2bf86cf)

For the moment, the display of HTML is too complex and it should be included when we will use CircleCI or ReadTheDocs.

bthirion reviewed (approved):
LGTM, thx.

lionelkusch commented:
@bthirion
The actual workflow add a comment in the PR for each testing commit. 
Do you think it's a good idea or it will create too much notification?

bthirion commented:
> @bthirion The actual workflow add a comment in the PR for each testing commit. Do you think it's a good idea or it will create too much notification?

Sorry, I'm not sure I get your point. Is this visible somewhere ?

lionelkusch commented:
> > @bthirion The actual workflow add a comment in the PR for each testing commit. Do you think it's a good idea or it will create too much notification?
> 
> Sorry, I'm not sure I get your point. Is this visible somewhere ?

see here: https://github.com/lionelkusch/hidimstat/pull/4

Remi-Gau commented:
> will create too much notification

in my experience this will create way too much noise

lionelkusch commented:
> my experience this will create way too much noise

Do you receive notification from CircleCi when the pipeline is run?

lionelkusch commented:
I am waiting the PR #214 for include the modification of the workflow.

lionelkusch commented:
![image](https://github.com/user-attachments/assets/1217761c-a214-44b6-bd0a-eaa41baa7c19)
I updated the pipeline of the CI. 
All the parts of the documentation are now on CircleCI.
I added the reports of the test in the documentation in order to have traces of them, especially about the speed of the tests.
I also include an option to skip the test when there is no requirement to run them.

Actually, there is an error in publishing the result because it is missing  access to a secret to pass the information to circleci. Once it is on main, it should be good.

Remi-Gau commented on .github/workflows/call_test_package.yml:
```diff
@@ -0,0 +1,107 @@
+      with:
+        python-version: ${{ matrix.python-version }}
+    
+    # Install dependency for running the tests
+    - name: Install the dependance
+      if: ${{ !inputs.skip_test && !startsWith(matrix.os, 'windows') }}
+      run: |
+        python -m venv .venv
+        source .venv/bin/activate
+        pip install --upgrade uv pip
+        uv pip install -e ".[test]"
+    - name: Install the dependance Windows
+      if: ${{ !inputs.skip_test && startsWith(matrix.os, 'windows') }}
+      run: |
+        python -m venv .venv
+        .venv\Scripts\python.exe -m pip install --upgrade uv pip
+        dir .venv\Scripts\
+        .venv\Scripts\uv.exe pip install -e ".[test]"
+    
+    # Run the tests with coverage and report add this end
+    # option for the coverage: --cov=. --cov-report=xml:... 
+    # option for mardown report: --md-report --md-report-output= ...
+    # option for html report: --self-contained-html --html=report-...
+    # option for duration report: --pytest-durations= number of second --pytest-resultlog=log-...
+    # inline command for being compatible on linux and windows
+    - name: Run pytest
+      if: ${{ !inputs.skip_test && !startsWith(matrix.os, 'windows') }}
+      run: |
+          source .venv/bin/activate
+          python -m pytest --cov=. --cov-report=xml:cov/cov-${{env.REPORT_NAME}}.xml --self-contained-html --html=report-${{env.REPORT_NAME}}.html --pytest-durations=10 --pytest-resultlog=log-${{env.REPORT_NAME}}.txt
+    - name: Run pytest  Windows
+      if: ${{ !inputs.skip_test && startsWith(matrix.os, 'windows') }}
+      run: |
+          .venv\Scripts/python.exe -m pytest --cov=. --cov-report=xml:cov/cov-${{env.REPORT_NAME}}.xml --self-contained-html --html=report-${{env.REPORT_NAME}}.html --pytest-durations=10 --pytest-resultlog=log-${{env.REPORT_NAME}}.txt
```
maybe in a follow up PR but you could this shorter by relying on tox to create the virtual environment and running the proper commands

this would also make it easier to run the same thing in CI and locally

lionelkusch commented:
I found a git action for setup uv: [here](https://docs.astral.sh/uv/guides/integration/github/)
This will reduce the complexity of the pipeline.
